## How to use this notebook
Any Jupyter Notebook (JN) is made of "text" (Markdown) and "code" cells. Code cells must be executed to see the result of the program. To run a cell, select it and press Shift + Enter. Using Shift + Enter multiple times will execute consecutive blocks of code one after another, while skipping text cells (executing them does nothing). It is important to run the code cells in the order they appear in the notebook.

A complete version of this JN is available by request to instructors using the book "Exploring mathematics with CAS assistance" for teaching. This version has blank or partially blank code lines that are supposed to be completed by the user before running the code.

Code cells contain (nonexecutable) comments preceded by the pound sign. The comments are of two types:
- a short comment placed on a code line typically states what the result of implementation of the encoded operation is
- a comment placed on a separate line either names the result of the following block of code lines or provides some details only for the next line where a more involved operation is encoded

This JN was written by Lydia Novozhilova.



# Lab 9: Submergence depth of a body of revolution

## Problem formulation
Consider a hollow body bounded by an ellipsoid of revolution (for example, a float for measuring fluid level) defined by the equation
$$\frac{x^2}{a^2} + \frac{y^2}{b^2} + \frac{z^2}{b^2} = 1.$$

Suppose that the body is partially submerged in a fluid of mass density $\rho_0.$ Let $\rho$ be the mean mass density of the body. Picture the submergence process as the rising level of fluid. Determine the depth $h$ of submerged part of the body assuming $h\in (-b,b).$

For more details refer to Section 8.3. **Suggestion**:
To simplify coding, rewrite the main equation (8.7) representing the Archimedes Principle as
$$ \frac{4c}{3}\pi ab^2 - V_0 = 0, \qquad (*)$$
where $V_0=V_0(h)$ is the volume of submerged part of the body and $c=\rho/\rho_0.$ Note that $0 < c < 1.$



##  Brief summary of this JN contents

Function **solve_main_eq** encodes the integrand for the slicing method and uses built-in Sympy function **integrate** to find the symbolic volume of submerged part of the ellipsoid (up to omitted factor $\pi a b$) as a function of the submergence depth $h$. (This task can be also done manually.) Then it constructs the main equation (*) for submergence depth and finds symbolic solutions of this equation.

Function **eval_sols** calls **solve_main_eq** function, substitutes numeric values for parameters $b$ and $c$ and returns three numeric solutions. It can be shown that only one of these solutions meets the requirement $h\in(-b,b)$ but this claim is not addressed here. The last code cell in this JN gives an example of how to encode selection of list elements satisfying certain condition.

The last section suggests a deeper investigation of the main equation for the height of submergence (not in the book). The investigation proves that for physically meaningful data the main equation always has three real roots.


In [ ]:
# loading Sympy library
from sympy import *


In [ ]:
# Constructing and solving the main equation.
# The factor pi*a*b present in the lhs of equation (*) is omitted.

def solve_main_eq():
  """
  Returns the cubic equation for the height of submergence and three roots of
  the equation in symbolic form
  """
  var('z h b c')
  integrd=1-?/? # encode integrand for implementing the slicing method
  expr = integrate(integrd,(z,-b,h))

  # Solving the main equation
  eq = ? # encode lhs of equation (*) with omitted factor pi*a*b
  print('Main equation for h: ',eq,' = 0')
  print()
  all_roots=solve(eq,h)
  # approximation of numeric values in roots
  return [N(s,4) for s in all_roots]


In [ ]:
# Test problem.
# The roots look scary, but with numberic parameters b, c they will look much better

solve_main_eq()

Main equation for h:  4*b*c/3 - 2*b/3 - h + h**3/(3*b**2)  = 0



[-0.7937*b**2/(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333 - 1.26*(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333,
 0.2646*b**2*(1.5 - 2.598*I)/(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333 + 3.78*(0.1667 + 0.2887*I)*(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333,
 0.2646*b**2*(1.5 + 2.598*I)/(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333 + 3.78*(0.1667 - 0.2887*I)*(b**3*c - 0.5*b**3 + (-0.25*b**6 + (b**3*c - 0.5*b**3)**2)**0.5)**0.3333]

In [ ]:
# Evaluation of symbolic solutions for numeric b and c.

def eval_sols(bval,cval):
  """
Args:
  bval, cval: values of parameters b, c (floats)
Output:
  List of numeric approximations of solutions to (*)
  """
  sym_sols=solve_main_eq()
  sols=[]
  for k in range(3):
    temp=sym_sols[k]
    # encode evaluation of parameters b and c
    temp1=temp.subs([(b,?),(c,?)])
    sol=temp1.evalf(3)
    sols.append(sol)
  print(sols)


In [ ]:
# Example 1.
# Comment: The extremely small imaginary parts of the roots below is an artifact
# stemming from numerical approximations and should be replaced with zeros.
eval_sols(2.,0.5)

Main equation for h:  4*b*c/3 - 2*b/3 - h + h**3/(3*b**2)
  
[-3.46 + 0.e-10*I, 3.46 - 0.e-11*I, 1.0e-16 - 1.11e-15*I]


In [ ]:
#Example 2.
eval_sols(0.5,0.3)

Main equation for h:  4*b*c/3 - 2*b/3 - h + h**3/(3*b**2)
  
[-0.79 + 0.e-9*I, 0.926 + 0.e-11*I, -0.137 + 0.e-9*I]


## Possible further exploration of the problem (not in the book)

The equation for depth of submersion can be written as a *depressed monic cubic* equation. We show that the three roots of this equation are real for all admissible values of the parameter $c$, $c\in(0,1)$. This is done in the last cell of the notebook.

A numerical experimentation with various $c$-values suggests that only one of the three real roots is physically meaningful, that is, lies in the interval $(-b,b).$ Rigorous analysis of this fact is possible, but it is beyond the scope of this lab. In the two examples above, the submergence depths are $0$ (meaning that exactly half of the float is submerged) and about $-0.137$, respectively. This selection of correct depth values is done by visual inspection of the solutions. If needed, a short snippet of code can make such selection automatically. The next cell gives an idea how to construct a sublist of elements satisfying some condition.

In [ ]:
# Useful code snippet: selecting list values that meet certain condition

l = [-3,1,-5,0,-1,2]
test = [n for n in l if (0 < n) & (n < 3)]
print(test)

[1, 2]


### Claim: For any $c\in(0,1)$ the main equation has three real roots
(This can be transformed into an additional advanced task of the lab)

The main equation $4bc/3 - 2b/3 - h + h^3/(3b^2)=0$ written as a monic cubic takes the form
$$h^3-3b^2h+2(2c-1)b^3=0.$$

Every depressed monic cubic equation $x^3+px+q=0$ defines a point $(p,q)$ in the $pq$-plane. A special curve, called **envelope,**  defined by the equation $4p^3+27q^2=0$ splits the $pq$-plane into two regions. As shown in Section 5.2, any depressed monic cubic equation represented by a point between the branches of the envelope has three real roots.

With a fixed parameter $c\in(0,1)$ and variable parameter $b>0,$ the main equation in this lab defines a parametric curve in the $pq$-plane
$$l(c): p=-3b^2,\ q=2(2c-1)b^3.$$

Corresponding explicit equation is
$$q=2(2c-1)\sqrt{-p^3/27}.$$
(Note that for $c=1/2$ the curve is just the negative part of the $p$-axis.)

Comparing this with the explicit equation of the envelope
$$q=\pm 2\sqrt{-p^3/27},$$
we conclude that for any $c\in(0,1),$ the curve $l(c)$ lies in the region between the branches of the envelope. Thus, the main equation always has three real roots.

The result of this analysis can be nicely visualized as a figure in $pq$-plane with the envelope graph and several curves with points representing the main equation for some interval of $b$-values and various $c$-values.

